### PR TITLE
Fixed addMachines.

### PR DESCRIPTION
**Pull request by petevg**
We now accept a single dict called params, which is what we get back
from the planner, rather than expecting expanded arguments.

We do some parsing of constraints, to put them in a format that should
make juju core happy.

@tvansteenburgh @johnsca: This is a partial fix for deploying hadoop processing. It gets us past all the errors when adding machines. We still get an error unmarshalling the instance.Placement object further down the line, but that error does not appear to stem from this part of the code (I'll work on fixing it next).

### CODE REVIEW

File: juju/constraints.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#
+# Library to parse constraints
```
Comment (tvansteenburgh) [inline juju/constraints.py]: s/Library/Module

File: juju/model.py [hunk 2]
```diff
@@ -1275,15 +1275,21 @@ def resolve(self, reference):
             case this machine is a container, in which case also ContainerType
             is set.
         """
```
Comment (tvansteenburgh) [inline juju/model.py]: This docstring needs to be updated.

File: juju/constraints.py (hunk 1 above)
Comment (petevg) [inline juju/constraints.py]: Done.

**Comment by petevg** [comment]
@tvansteenburgh I agree with your assessment that #10 wasn't a real bug (or rather, it had a real bug initially, where it was accepting a "params" argument, then clobbering it by saying "params=dict()", but @johnsca fixed that earlier).

That said, there is a bug here, which I filed with core: the client shouldn't be responsible for parsing the constraints. constraints.py in this PR is a hopefully short lived workaround, with a future version of the Facade eliminating the need for it.

**Comment by petevg** [comment]
@tvansteenburgh nits addressed. Also added a link to the bug that I filed w/ juju core, along with a notice that constraints.py should eventually go away. What do you think?

File: juju/model.py [hunk 3]
```diff
@@ -1255,35 +1256,38 @@ def resolve(self, reference):
...
+    async def addMachines(self, params):
+        """
+        :param params dict:
+            Dictionary specifying the machine to add. Keys include:
+
+            series: string specifying the machine OS series.
+            constraints: string holding optional machine constraints. We'll
+                parse this into the json friendly dict that the juju api
+                expects.
+            Container_type: string holding the type of the container (for
```
Comment (tvansteenburgh) [inline juju/model.py]: Is this really Container_type and not container_type?

File: juju/model.py [hunk 4]
```diff
@@ -1255,35 +1256,38 @@ def resolve(self, reference):
...
-            container_type=container_type,
-            parent_id=self.resolve(parent_id),
-        )
-        results = await self.client_facade.AddMachines(params)
-        log.debug('Added new machine %s', results[0].machine)
-        return results[0].machine
+    async def addMachines(self, params):
+        """
+        :param params dict:
+            Dictionary specifying the machine to add. Keys include:
```
Comment (tvansteenburgh) [inline juju/model.py]: Are all of the keys in `params` optional? Would be helpful to say which, if any, are required.

Comment (petevg) [inline juju/model.py]: @tvansteenburgh Added a note to the effect that all params are optional (as far as I can tell).

Also made the params dict optional, though I think that you'd only call this method if you had some params to specify.

File: juju/model.py [hunk 5]
```diff
@@ -1255,35 +1256,41 @@ def resolve(self, reference):
...
+        """
+        :param params dict:
+            Dictionary specifying the machine to add. All keys are optional.
+            Keys include:
+
+            series: string specifying the machine OS series.
+            constraints: string holding machine constraints, if any. We'll
+                parse this into the json friendly dict that the juju api
+                expects.
+            Container_type: string holding the type of the container (for
```
Comment (tvansteenburgh) [inline juju/model.py]: Should this be `container_type` instead?

Comment (petevg) [inline juju/model.py]: @tvansteenburgh Yep. That's what I get for copy-pasting without double checking to make sure the original author was right. Fixed. :-)

**Pull request merged**